src/main/scala/com/fulcrumgenomics/commons/util/ConfigurationLike.scala

```diff
@@ -0,0 +1,247 @@
+  *
+  * @example Retrieve a value at a given path, where the path must exist:
+  * {{{
+  *   scala> import com.fulcrumgenomics.commons.util.ConfigurationLike
+  *   scala> import com.typesafe.config.{Config, ConfigFactory}
+  *   scala> val conf = new ConfigurationLike {
+  *     val config: Config = ConfigFactory.parseString("""a.str = "string", a.int = 2, a.set = [1,2,3]""")
+  *   }
+  *   scala> conf[String]("a.str")
+  *   res0: String = string
+  *   scala> conf[Int]("a.int")
+  *   res1: Int = 2
+  *   scala> conf[Set[Int]]("a.set")
+  *   res2: Set[Int] = Set(1, 2, 3)
+  *   scala> conf[Set[Int]]("a.does-not-exist")
+  *   java.lang.IllegalArgumentException: Exception retrieving configuration key 'a.does-not-exist': No configuration setting found for key 'a.does-not-exist'
+  *   ...
+  * }}}
+  *
+  * @example Retrieve a value at a given path, where the path may not exist
+  * {{{
+  *   scala> import com.fulcrumgenomics.commons.util.ConfigurationLike
+  *   scala> import com.typesafe.config.{Config, ConfigFactory}
+  *   scala> val conf = new ConfigurationLike {
+  *     val config: Config = ConfigFactory.parseString("""a.str = "string", a.int = 2, a.set = [1,2,3]""")
+  *   }
+  *   scala> conf.get[String]("a.str")
+  *   res0: Option[String] = Some(string)
+  *   scala> conf.get[Int]("a.int")
+  *   res1: Option[Int] = Some(2)
+  *   scala> conf.get[Set[Int]]("a.set")
+  *   res2: Option[Set[Int]] = Some(Set(1, 2, 3))
+  *   scala> conf.get[Set[Int]]("a.does-not-exist")
+  *   res3: Option[Set[Int]] = None
+  * }}}
+  *
+  * @example Retrieve a value at a given path if it exists, otherwise return a default:
+  * {{{
+  *   scala> import com.fulcrumgenomics.commons.util.ConfigurationLike
+  *   scala> import com.typesafe.config.{Config, ConfigFactory}
+  *   scala> val conf = new ConfigurationLike {
+  *     val config: Config = ConfigFactory.parseString("""a.str = "string", a.int = 2, a.set = [1,2,3]""")
+  *   }
+  *   scala> conf.getOrElse[String]("a.does-not-exist", "default")
+  *   res0: String = default
+  *   scala> conf.getOrElse[Int]("a.does-not-exist", 0)
+  *   res1: Int = 0
+  *   scala> conf.getOrElse[Set[Int]]("a.does-not-exist", Set.empty)
+  *   res2: Set[Int] = Set()
+  * }}}
+  */
+trait ConfigurationLike {
+  /** The configuration to query. */
+  protected def config : Config
+
+  /** A logger to which to report exceptions. */
+  protected def _logger: Option[Logger] = None
+
+  /** The keys requested, in order. */
+  private val _requestedKeys = new java.util.concurrent.ConcurrentSkipListSet[String]()
+
+  /**
+    * Looks up a single value of a specific type in configuration. If the configuration key
+    * does not exist, an exception is thrown. If the requested type is not supported, an
+    * exception is thrown.
+    */
+  def apply[T : TypeTag](path: String) : T = {
+    this._requestedKeys.add(path)
+    try {
+      asType[T](path)
+    }
+    catch {
+      case ex : Exception =>
```

Comment:
A couple of things here:

1. I think exceptions should _either_ be logged or re-thrown but not both as that will almost always lead to double-logging which can be confusing.  I tend to think that low-level APIs like this should _never_ log and only communicate by succeeding or throwing exceptions.
2. Similar problem with embedding the prior exception's message and also chaining it.  It will lead to some part of the information being printed twice which is more confusing than helpful.  I would get rid of the `: ${ex.getMessage}` and just chain it.